### PR TITLE
Introduce gosec as Static Code Analysis tool

### LINKED ISSUE #1712: Add gosec to the ci pipeline
Adding another static analyzer should be easy and will save us from audits

### PR DESCRIPTION
**Pull request by yuvalk**
I've added gosec to the CI job
then went over the results and fixed all the high/medium issues.

for now, I've left LOW out of scanner result as this seems to include too many false positives and the benefit from investing in these is .. low..

this fixes #1712

### CODE REVIEW

File: configmaptocrs/main.go [hunk 1]
```diff
@@ -105,7 +107,11 @@ func generate(w io.Writer, origin string) error {
 
 func readConfig(origin string) ([]byte, error) {
 	fp := filepath.Join(inputDirPath, origin)
-	f, err := os.Open(fp)
+	fp = filepath.Clean(fp)
+	if !strings.HasPrefix(fp, path.Clean(inputDirPath)) {
```
Comment (fedepaol) [inline configmaptocrs/main.go:115]: we do we need this? fp is the join of inputdir and origin, so it will have inputDirPath as prefix

File: configmaptocrs/main.go [hunk 2]
```diff
@@ -9,8 +9,10 @@ import (
 	"io/ioutil"
 	"log"
 	"os"
+	"path"
```
Comment (fedepaol) [inline configmaptocrs/main.go:12]: What is this commit fixing? Or, what is G304 ?

File: frr-metrics/exporter.go [hunk 3]
```diff
@@ -8,6 +8,7 @@ import (
 	stdlog "log"
 	"net/http"
 	"os"
+	"time"
```
Comment (fedepaol) [inline frr-metrics/exporter.go:11]: Why is this required  / raised by gosec?

File: internal/config/config.go [hunk 4]
```diff
@@ -384,6 +384,7 @@ func peerFromCR(p metallbv1beta2.BGPPeer, passwordSecrets map[string]corev1.Secr
 
 	var nodeSels []labels.Selector
 	for _, s := range p.Spec.NodeSelectors {
+		s := s // so we can use &s
```
Comment (fedepaol) [inline internal/config/config.go:387]: Neat. This is not causing bugs as we are using &s in pure functions and not carrying it around, but might bite us in the future.

**Comment by fedepaol** [comment]
Thanks for the contribution! Left a few requests for clarification.

File: frr-metrics/exporter.go (hunk 3 above)
Comment (yuvalk) [reply frr-metrics/exporter.go:11]: see below, without a ReadTimeout, the defulat is infinite,
which make it easy for denial-of-service (DoS) attacks

File: configmaptocrs/main.go (hunk 2 above)
Comment (yuvalk) [reply configmaptocrs/main.go:12]: I can add a better commit msg
https://securego.io/docs/rules/g304.html

File: configmaptocrs/main.go (hunk 1 above)
Comment (yuvalk) [reply configmaptocrs/main.go:115]: actually no
I had to do it to workaround https://github.com/securego/gosec/issues/893
but still, IMHO, is a better option than masking it (with a //#nosec comment)

Comment (yuvalk) [reply configmaptocrs/main.go:115]: again, I can add a better comment/commit msg so it'll be properly documented

File: configmaptocrs/main.go [hunk 5]
```diff
@@ -105,7 +107,11 @@ func generate(w io.Writer, origin string) error {
 
 func readConfig(origin string) ([]byte, error) {
 	fp := filepath.Join(inputDirPath, origin)
-	f, err := os.Open(fp)
+	fp = filepath.Clean(fp)
+	if !strings.HasPrefix(fp, path.Clean(inputDirPath)) {
+		return nil, fmt.Errorf("Unsafe path %s", origin)
+	}
+	f, err := os.Open(filepath.Clean(fp))
```
[the remark this answers is not in the document]
Comment (yuvalk) [reply configmaptocrs/main.go:118]: actually no
I had to do it to workaround https://github.com/securego/gosec/issues/893
but still, IMHO, is a better option than masking it (with a //#nosec comment)

[the remark this answers is not in the document]
Comment (yuvalk) [reply configmaptocrs/main.go:118]: again, I can add a better comment/commit msg so it'll be properly documented

File: configmaptocrs/main.go (hunk 1 above)
Comment (yuvalk) [reply configmaptocrs/main.go:115]: no, consider the case
inputDirPath = "/var/input"
and 
origin = "../../etc/passwd"

the join would be "/var/input/../../etc/passwd"
and the clean is "/etc/passwd" 
which doesnt have the inputDirPath prefix.

for (hopefully) obvious reasons, we want to block such cases.

**Comment by fedepaol** [comment]
Would you mind fixing the conflicts?

File: internal/config/config.go (hunk 4 above)
Comment (yuvalk) [reply internal/config/config.go:387]: yeah, reinforcing the code :-)
could argue the other way too (adding bullshit, making unnecessary copy, etc etc)
but it's the easiest way for the code scanner to see there are no issues here

**Comment by yuvalk** [comment]
> Would you mind fixing the conflicts?

I think this is all set now

**Comment by fedepaol** [comment]
LGTM, thanks!
